circuits/circuits/claiming/challenge.circom

```diff
@@ -0,0 +1,144 @@
+template FilteredLeafIsValid(claim_levels, state_levels, upper_state_levels, i) {
+    // Merkle roots
+    signal input states_history_root;
+    signal input claim_root;
+    signal input filtered_root;
+
+    // Defender
+    signal input defender[3];
+    signal input defender_state_PathElements[state_levels];
+    signal input defender_state_PathIndices[state_levels];
+    signal input defender_claim_PathElements[claim_levels];
+
+    // Challenger
+    signal input challenger_index;
+    signal input challenger[3];
+    signal input challenger_state_PathElements[state_levels];
+    signal input challenger_state_PathIndices[state_levels];
+    signal input challenger_claim_PathElements[claim_levels];
+
+    // Results
+    signal input filtered_pathElements[claim_levels];
+
+    signal claim_is_valid <== ClaimIsValid(claim_levels, state_levels)(
+        i,
+        defender,
+        defender_state_PathElements,
+        defender_state_PathIndices,
+        defender_claim_PathElements,
+        states_history_root,
+        claim_root
+    );
+
+    // Make sure the challenger is a real leaf in the claim tree with a valid proof of ownership
+    // The prover can always find *some* challenger in the tree as long as there is at least one valid claim
+    signal challenge_is_valid <== ClaimIsValid(claim_levels, state_levels)(
+        challenger_index,
+        challenger,
+        challenger_state_PathElements,
+        challenger_state_PathIndices,
+        challenger_claim_PathElements,
+        states_history_root,
+        claim_root
+    );
+    challenge_is_valid === 1;
+
+    signal challenge_succeeds <== ChallengeSucceeds(state_levels, upper_state_levels)(
+        defender <== defender,
+        challenger <== challenger,
+        defender_state_indices <== defender_state_PathIndices,
+        challenger_state_indices <== challenger_state_PathIndices
+    );
+
+    // if the claim is valid and the challenge doesn't succeed, the leaves in the claim and filtered trees should be equal
+    signal should_be_equal <== claim_is_valid * (1 - challenge_succeeds);
+    signal claim_hash <== Poseidon(3)(defender);
+    signal filtered_leaf <== claim_hash * should_be_equal;
+    CheckMerkleProofStrict(claim_levels)(
+        filtered_leaf,
+        filtered_pathElements,
+        Num2Bits(claim_levels)(in <== i),
+        filtered_root
+    );
+}
+
+// Asserts that the claim is in the claim tree
+// Returns whether the claim is properly formed and is in the state history
+template ClaimIsValid(claim_levels, state_levels) {
+    signal input i;
+    signal input claim[3];
+    signal input state_PathElements[state_levels];
+    signal input state_PathIndices[state_levels];
+    signal input claim_PathElements[claim_levels];
+    signal input states_history_root;
+    signal input claim_root;
+
+    signal output out;
+
+    // Ensure that claim was made at some point during the claiming period
```

Comment:
I think this is important for making the final tree deterministic wrt the claims. And it's important to make it deterministic so that people can easily generate merkle proofs to get their money out of the final tree.

For example, if the prover adds a new valid claim that supersedes another claim that would have won, the prover will be the only person who can generate merkle proofs. No one else will be able to figure out what the new valid claim is without a lot of luck. The prover could then hold the coins ransom.